src/goDeclaration.ts

```diff
@@ -24,216 +24,230 @@ export interface GoDefinitionInformation {
-					let doc = '';
-					let sigName = signature.substring(0, signature.indexOf(' '));
-					let sigParams = signature.substring(signature.indexOf(' func') + 5);
-					let searchSignature = 'func ' + sigName + sigParams;
-					for (let i = 0; i < godocLines.length; i++) {
-						if (godocLines[i] === searchSignature) {
-							while (godocLines[++i].startsWith('    ')) {
-								doc += godocLines[i].substring(4) + '\n';
-							}
-							break;
-						}
+					callback = this.gogetdocCallback;
+					break;
+				default:
+					break;
+			}
+
+			if (!args || !callback) {
+				return Promise.resolve(null);
+			}
+
+			this.cancelCurrentRunningProcess();
+			return new Promise<GoDefinitionInformation>((resolve, reject) => {
+				this.currentRunningProcess = cp.execFile(binaryPath, args, { env }, (err, stdout, stderr) => {
+					if (err) {
+						return reject(err);
 					}
-					if (doc !== '') {
-						definitionInformation.doc = doc;
+
+					let definitionInformation: GoDefinitionInformation;
+					try {
+						definitionInformation = callback(stdout);
+					} catch (e) {
+						return reject(e);
+					}
+
+					if (tool !== 'godef' || !includeDocs || !definitionInformation.declarationlines || !definitionInformation.declarationlines.length) {
+						return resolve(definitionInformation);
 					}
-					return resolve(definitionInformation);
+
+					// godef doesnt provide docs, so we call godoc
+					let pkgPath = path.dirname(definitionInformation.file);
+					let signature = definitionInformation.declarationlines[0];
+					let godoc = getBinPath('godoc');
+					this.currentRunningProcess = cp.execFile(godoc, [pkgPath], { env }, (err, stdout, stderr) => {
+						if (err && (<any>err).code === 'ENOENT') {
+							vscode.window.showInformationMessage('The "godoc" command is not available.');
+							return resolve(definitionInformation);
+						}
+						try {
+							let godocLines = stdout.toString().split('\n');
+							let doc = '';
+							let sigName = signature.substring(0, signature.indexOf(' '));
+							let sigParams = signature.substring(signature.indexOf(' func') + 5);
+							let searchSignature = 'func ' + sigName + sigParams;
+							for (let i = 0; i < godocLines.length; i++) {
+								if (godocLines[i] === searchSignature) {
+									while (godocLines[++i].startsWith('    ')) {
+										doc += godocLines[i].substring(4) + '\n';
+									}
+									break;
+								}
+							}
+							if (doc !== '') {
+								definitionInformation.doc = doc;
+							}
+						} catch (e) {
+							return reject(e);
+						}
+						return resolve(definitionInformation);
+					});
+					this.currentRunningProcess.on('close', () => { this.currentRunningProcess = null; });
 				});
-			} catch (e) {
-				reject(e);
-			}
+				this.currentRunningProcess.stdin.end(tool === 'godef' ? document.getText() : getFileArchive(document));
+				this.currentRunningProcess.on('close', () => { this.currentRunningProcess = null; });
```

Comment:
I think the nulling out should be done in `cancelCurrentRunningProcess()`. I think there could be a race condition here. If this happens

- cancelCurrentRunningProcess() kills the process
- 'close' event is queued but not handled yet
- currentRunningProcess is immediately assigned to a newly spawned process
- 'close' handler runs and nulls out the prop, but the process is actually still running.

Reply:
Done

Reply:
Done